core-metastore/src/main/java/org/zalando/nakadi/repository/db/SubscriptionDbRepository.java

```diff
@@ -152,24 +157,59 @@ public List<Subscription> listSubscriptions(final Set<String> eventTypes, final
             params.add(pp.offset);
         });
 
-        try {
-            return jdbcTemplate.query(queryBuilder.toString(), params.toArray(), rowMapper);
-        } catch (final DataAccessException e) {
-            LOG.error("Database error when listing subscriptions", e);
-            throw new ServiceTemporarilyUnavailableException("Error occurred when running database request");
+        return executeListSubscriptions(queryBuilder.toString(), params);
+    }
+
+    /**
+     * Method lists subscriptions.
+     *
+     * @param eventTypes        Event types that are targeted by subscription
+     * @param owningApplication Ownnig application of subscriptions
+     * @param reader            Authorization attribute of a reader of subscripitons
+     * @param token             Token to use for iteration
+     * @param limit             Amount of records to return. Min value is 1
+     * @return List of subscriptions with optional tokens to iterate forward and backward
+     */
+    public ListResult listSubscriptions(
+            final Set<String> eventTypes,
+            final Optional<String> owningApplication,
+            final Optional<AuthorizationAttribute> reader,
+            @Nullable final Token token,
+            final int limit) {
+        if (limit < 1) {
+            throw new IllegalArgumentException("Limit can't be less than 1");
+        }
+
+        final StringBuilder queryBuilder = new StringBuilder();
+        final List<String> clauses = Lists.newArrayList();
+        final List<Object> params = Lists.newArrayList();
+
+        applyFilter(eventTypes, owningApplication, reader, queryBuilder, clauses, params);
+
+        if (null == token || !token.hasSubscriptionId()) {
+            return listNoToken(queryBuilder, clauses, params, limit);
+        } else if (token.getTokenType().forward) {
+            return listForwardToken(queryBuilder, clauses, params, token, limit);
+        } else {
+            return listBackwardToken(queryBuilder, clauses, params, token, limit);
         }
     }
 
-    static void applyFilter(
+    private static void applyFilter(
             final Set<String> eventTypes,
             final Optional<String> owningApplication,
             final Optional<AuthorizationAttribute> reader,
+            final StringBuilder queryBuilder,
             final List<String> clauses,
             final List<Object> params) {
+
+        queryBuilder.append("SELECT s_subscription_object FROM zn_data.subscription");
```

Comment:
With this approach isn't QueryBuilder a return result of the function?

Reply:
Yes, we can rename it to `initQuery` or something and return `StringBuilder` from it.  The whole business with buliding the query from blocks here is quite ugly anyway.